oeps/oep-0036-microservice-routing-layer.rst

```diff
@@ -0,0 +1,93 @@
+===================================
+OEP-36: Microservice Routing Layer
+===================================
+
++---------------+-------------------------------------------+
+| OEP           | :doc:`oep-0036-microservice-routing-layer`|
++---------------+-------------------------------------------+
+| Title         | Microservice Routing Layer                |
++---------------+-------------------------------------------+
+| Last-Modified | 2019-04-25                                |
++---------------+-------------------------------------------+
+| Author        | Fred Smith <derf@edx.org>           |
++---------------+-------------------------------------------+
+| Arbiter       |                |
++---------------+-------------------------------------------+
+| Status        | Provisional                               |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2019-04-25                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+Proposes a system with which users can interact with microservice applications
+using a layer 7 routing framework. 
+
+Motivation
+==========
+
+Microservices are currently implemented using layer 4 (DNS) based routing. Each
+microservice occupies a unique hostname, with dedicated ELB or CDN 
+infrastructure to support routing user traffic to the service. This pattern
+is simple to monitor, deploy, and manage, but slow and expensive to build,
+and regularly causes problems with cross service user information sharing. In
+Addition, modifying or splitting a monolithic service into microservices resuls
+in confusing URL changes and costly migrations.
+
+A layer 7 routing layer would allow us to separate the user facing URL or API
+endpoint from the backend implementation. It would allow microservices to exist
+that only manage a small number of endpoints, without requring developers to 
+build out load balancing or CDN infrastructure. 
+
+We are currently using a layer 7 routing framework for prospectus, allowing
+us to replace drupal generated pages with gatsby generated pages quickly and
+seamlessly.
+
+This change will allow developers, openedx operators and resellers to deploy
+new openedx sites quicker and easier than they currently can.  The router
+will provide a map for people to understand which services they need to deploy.
```

Comment:
Some thoughts and questions: 

Will the router handle auth in any way or just routing?

Is the router a required component to run our stack now? (because it does stuff with auth)

It is important to note that if the router is a required component that it definitely cannot be k8 native.

If it is not a required component we can use a k8s native one like ambassador, but it will not provide this 'map for people to understand which services they need to deploy' at that point.

Do we configure the services to talk to one another with or without knowledge of the router?

If we want to look at how services discover one another as well as how services are routed from the outside world we might consider a service mesh and a L7 router.

Reply:
I've addressed the capabilities questions, in that, it might.  We should figure out if auth is something we want from this component.  We have this capability with our current platform (nginx), as well as request throttling, and have typically pushed that role up the stack to django.

The routing layer is a required component.

I would suggest that all inter-service communication go through this routing layer, to minimize areas of configuration and access control.